shell/utils/dynamic-content/index.ts

```diff
@@ -0,0 +1,273 @@
+ * This timeout is set relatively low (15s). The default, otherwise, is 2 minutes.
+ *
+ * @param timeoutMs Time in milliseconds after which the abort signal should signal
+ */
+function newRequestAbortSignal(timeoutMs: number) {
+  const abortController = new AbortController();
+
+  setTimeout(() => abortController.abort(), timeoutMs || 0);
+
+  return abortController.signal;
+}
+
+/**
+ * Update the local storage data that tracks when to next fetch content and how many consecutive errors we have had
+ *
+ * @param didError Indicates if we should update to record content retrieved without error or with error
+ */
+function updateFetchInfo(didError: boolean) {
+  if (!didError) {
+    // No error, so check again tomorrow and remove the backoff setting, so it will get its default next time
+    const nextFetch = day().add(1, 'day');
+    const nextFetchString = nextFetch.format(UPDATE_DATE_FORMAT);
+
+    window.localStorage.setItem(LOCAL_STORAGE_UPDATE_FETCH_DATE, nextFetchString);
+    window.localStorage.removeItem(LOCAL_STORAGE_UPDATE_ERRORS);
+  } else {
+    // Did error, read the backoff, increase and add to the date
+    const contiguousErrorsString = window.localStorage.getItem(LOCAL_STORAGE_UPDATE_ERRORS) || '0';
+
+    let contiguousErrors = parseInt(contiguousErrorsString, 10);
+
+    // Increase the number of errors that have happened in a row
+    contiguousErrors++;
+
+    // Once we reach the max backoff, just stick with it
+    if (contiguousErrors >= BACKOFFS.length ) {
+      contiguousErrors = BACKOFFS.length - 1;
+    }
+
+    // Now find the backoff (days) given the error count and calculate the date of the next fetch
+    const daysToAdd = BACKOFFS[contiguousErrors];
+    const nextFetch = day().add(daysToAdd, 'day');
+    const nextFetchString = nextFetch.format(UPDATE_DATE_FORMAT);
+
+    window.localStorage.setItem(LOCAL_STORAGE_UPDATE_FETCH_DATE, nextFetchString);
+    window.localStorage.setItem(LOCAL_STORAGE_UPDATE_ERRORS, contiguousErrors.toString());
+  }
+}
+
+/**
+ * Fetch dynamic content (if needed)
+ */
+export async function fetchDynamicContent(context: Context): Promise<Partial<DynamicContent> | undefined> {
+  const { getters, logger, config } = context;
+
+  // Check if we already have done an update check today
+  let content: Partial<DynamicContent> = {};
+
+  try {
+    const today = day();
+    const todayString = today.format(UPDATE_DATE_FORMAT);
+    const nextFetch = window.localStorage.getItem(LOCAL_STORAGE_UPDATE_FETCH_DATE) || todayString;
+
+    // Read the cached content from local storage if possible
+    content = JSON.parse(window.localStorage.getItem(LOCAL_STORAGE_UPDATE_CONTENT) || '{}');
+
+    const nextFetchDay = day(nextFetch);
+
+    // Just in case next day gets reset to the past or corrupt, otherwise next fetch needs to not be in the future
+    if (!nextFetchDay.isValid() || !nextFetchDay.isAfter(today)) {
+      logger.info(`Performing update check on ${ todayString }`);
```

Comment:
i think i understand why there's the same log at both levels (to read the flow of events when filtering logs by verbose)?

just wondering if there's value in another approach...

1. shell/utils/dynamic-content/util.ts `log` prefixes all dynamic content messages with some kind of prefix - `Dynamic Content: ` + message
2. in browser console filtering by `Dynamic Content: ` will then show all applicable output

Is this a pattern... log lines requiring prefix with context... we could consider optionally applying everywhere? it would help around the growing pile of dev-only warning spam